crypto/pkcs7/pkcs7.c

```diff
@@ -1409,3 +1415,246 @@ int PKCS7_decrypt(PKCS7 *p7, EVP_PKEY *pkey, X509 *cert, BIO *data, int flags) {
+      goto err;
+    }
+    if (!BIO_get_md_ctx(bio, &mdc) || !mdc) {
+      OPENSSL_PUT_ERROR(PKCS7, ERR_R_PKCS7_LIB);
+      goto err;
+    }
+    if (EVP_MD_CTX_type(mdc) == md_type) {  // found it!
+      break;
+    }
+    bio = BIO_next(bio);
+  }
+
+  // Make a copy of |mdc| so we don't finalize the MD_CTX owned by |bio|
+  if ((mdc_copy = EVP_MD_CTX_new()) == NULL ||
+      !EVP_MD_CTX_copy_ex(mdc_copy, mdc)) {
+    OPENSSL_PUT_ERROR(PKCS7, ERR_R_EVP_LIB);
+    goto err;
+  }
+
+  // We don't currently support signed attributes. See |PKCS7_NOATTR|.
+  if (si->auth_attr && sk_X509_ATTRIBUTE_num(si->auth_attr) != 0) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_INVALID_SIGNED_DATA_TYPE);
+    goto err;
+  }
+
+  EVP_PKEY *pkey;
+  if ((pkey = X509_get0_pubkey(signer)) == NULL) {
+    goto err;
+  }
+
+  ASN1_OCTET_STRING *data_body = si->enc_digest;
+  if (!EVP_VerifyFinal(mdc_copy, data_body->data, data_body->length, pkey)) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_SIGNATURE_FAILURE);
+    goto err;
+  }
+
+  ret = 1;
+
+err:
+  EVP_MD_CTX_free(mdc_copy);
+  return ret;
+}
+
+int PKCS7_verify(PKCS7 *p7, STACK_OF(X509) *certs, X509_STORE *store,
+                 BIO *indata, BIO *out, int flags) {
+  GUARD_PTR(p7);
+  GUARD_PTR(store);
+  STACK_OF(X509) *signers = NULL, *untrusted = NULL;
+  X509_STORE_CTX *cert_ctx = NULL;
+  BIO *p7bio = NULL;
+  int ret = 0;
+
+  if (!PKCS7_type_is_signed(p7)) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_WRONG_CONTENT_TYPE);
+    goto out;
+  }
+
+  // If |p7| is detached, caller must supply data via |indata|
+  if (PKCS7_is_detached(p7) && indata == NULL) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_NO_CONTENT);
+    goto out;
+  }
+
+  // We enforce OpenSSL's PKCS7_NO_DUAL_CONTENT flag in all cases for signed
+  if (PKCS7_type_is_signed(p7) && (!PKCS7_is_detached(p7) && indata)) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_CONTENT_AND_DATA_PRESENT);
+    goto out;
+  }
+
+  STACK_OF(PKCS7_SIGNER_INFO) *sinfos = PKCS7_get_signer_info(p7);
+  if (sinfos == NULL || sk_PKCS7_SIGNER_INFO_num(sinfos) == 0UL) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_NO_SIGNATURES_ON_DATA);
+    goto out;
+  }
+
+  if ((signers = pkcs7_get0_signers(p7, certs, flags)) == NULL) {
+    OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_NO_SIGNERS);
+    goto out;
+  }
+
+  if (!(flags & PKCS7_NOVERIFY)) {
+    STACK_OF(X509) *included_certs = pkcs7_get0_certificates(p7);
+    if ((cert_ctx = X509_STORE_CTX_new()) == NULL ||
+        !pkcs7_x509_add_certs_new(&untrusted, certs) ||
+        !pkcs7_x509_add_certs_new(&untrusted, included_certs)) {
+      OPENSSL_PUT_ERROR(PKCS7, ERR_R_PKCS7_LIB);
+      goto out;
+    }
+
+    for (size_t k = 0; k < sk_X509_num(signers); k++) {
+      X509 *signer = sk_X509_value(signers, k);
+      if (!X509_STORE_CTX_init(cert_ctx, store, signer, untrusted)) {
+        OPENSSL_PUT_ERROR(PKCS7, ERR_R_X509_LIB);
+        goto out;
+      }
+      if (!X509_STORE_CTX_set_default(cert_ctx, "smime_sign")) {
```

Comment:
Should `"smime_sign"` be assigned to a macro or constant somewhere? (I see the same raw string is likewise used [here](https://github.com/aws/aws-lc/blob/main/crypto/x509/x509_vpm.c#L477).)

Reply:
i don't think so. it's documented [here](https://github.com/aws/aws-lc/blob/8226a0565dd81aab1c3eaf4e623301b7ce2c4423/include/openssl/x509.h#L2805) and pulling it up into a shared macro would break the cross-module abstraction between `pkcs7` and `x509`. we're just a regular caller here.